python/lsst/ip/isr/brighterFatterKernel.py

```diff
@@ -0,0 +1,413 @@
+            self.badAmps = []
+
+            for amp in detector:
+                ampName = amp.getName()
+                self.means[ampName] = []
+                self.variances[ampName] = []
+                self.rawXcorrs[ampName] = []
+                self.gain[ampName] = amp.getGain()
+                self.noise[ampName] = amp.getReadNoise()
+                self.meanXcorrs[ampName] = []
+                self.ampKernels[ampName] = []
+                self.valid[ampName] = []
+        elif self.level == 'DETECTOR':
+            for det in camera:
+                detName = det.getName()
+                self.detKernels[detName] = []
+
+        return self
+
+    def getLengths(self):
+        """Return the set of lengths needed for reshaping components.
+
+        Returns
+        -------
+        kernelLength : `int`
+            Product of the elements of self.shape.
+        smallLength : `int`
+            Size of an untiled covariance.
+        nObs : `int`
+            Number of observation pairs used in the kernel.
+        """
+        kernelLength = self.shape[0] * self.shape[1]
+        smallLength = int((self.shape[0] - 1)*(self.shape[1] - 1)/4)
+        nObservations = set([len(self.means[amp]) for amp in self.means])
+        if len(nObservations) != 1:
+            raise RuntimeError("Inconsistent number of observations found.")
+        nObs = nObservations.pop()
+
+        return (kernelLength, smallLength, nObs)
+
+    @classmethod
+    def fromDict(cls, dictionary):
+        """Construct a calibration from a dictionary of properties.
+
+        Parameters
+        ----------
+        dictionary : `dict`
+            Dictionary of properties.
+
+        Returns
+        -------
+        calib : `lsst.ip.isr.BrighterFatterKernel
+            Constructed calibration.
+
+        Raises
+        ------
+        RuntimeError :
+            Raised if the supplied dictionary is for a different
+            calibration.
+        """
+        calib = cls()
+
+        if calib._OBSTYPE != (found := dictionary['metadata']['OBSTYPE']):
+            raise RuntimeError(f"Incorrect brighter-fatter kernel supplied.  Expected {calib._OBSTYPE}, "
+                               f"found {found}")
+
+        calib.setMetadata(dictionary['metadata'])
+        calib.calibInfoFromDict(dictionary)
+
+        calib.level = dictionary['metadata'].get('LEVEL', 'AMP')
+        calib.shape = (dictionary['metadata'].get('KERNEL_DX', 0),
+                       dictionary['metadata'].get('KERNEL_DY', 0))
+        calib.badAmps = dictionary['badAmps']
+
+        calib.means = {amp: np.array(dictionary['means'][amp]) for amp in dictionary['means']}
+        calib.variances = {amp: np.array(dictionary['variances'][amp]) for amp in dictionary['variances']}
+
+        # Lengths for reshape:
+        _, smallShape, nObs = calib.getLengths()
+
+        calib.rawXcorrs = {amp: np.array(dictionary['rawXcorrs'][amp]).reshape((nObs,
```

Comment:
I'm still a bit confused by the reshaping that is happening here. If, let's say, the original shape is `(8,8)` (default) and `nObs` is, say, `20`, then `rawXcorrs` total length is `20*8*8 = 1280`. However, `smallShape` here (which I think has a more appropriate name inside the function, as it is `smallLength`) would be `int((8-1)*(8-1)/4) = 12`, so the reshape here would be `1280-->2800`? I'm probably missing something, sorry!

Reply:
And, why the reshape pf the covariance matrices from PTC? in preparation for the tiling when doing the convolution?

Reply:
The `calib.shape` is the shape of the kernel, which has been tiled to fill all quadrants.  As in the previous comment, this is 17x17 for an 8x8 covariance.  So for 20 observations, the returned lengths will be `kernelLength = 17*17 = 289, smallLength = ((17-1)*(17-1))/4 = 8*8 = 64, nObs = 20`.

The reshape is here just because the code stores the covariances for future validation, and they need to be persistable to FITS.

However, in looking at this block again, I made a mistake on the code deduplication ticket.  As you point out, `smallLength` is the better name, as it's just a single value.  However, this is the "to FITS" reshape size, so in constructing the calibration ("from FITS"), it needs to use the `sqrt` of this value as the shapes.

I think this also implies I need to update the unit tests as well, as a read/write test would have found that error.